### PR TITLE
DOC: Update documentation for user-defined converters in np.genfromtxt

### LINKED ISSUE #23429: DOC: User defined Converters in `np.genfromtxt` get extraneous inputs which the user is not told about.
### Issue with current documentation:

Converters are functions that the user defines and are passed to `np.genfromtxt` to process particular columns.

A user who wants to ensure that the data in a column exists only within a small subset of possible values might want a converter to throw an error if it reads an entry with a bad value. For example, a user might want to ensure that each entry in a column is either `'A'` or `'B'`, and raise an exception otherwise.

This is problematic in terms of how it interacts with some of the code internal to `np.genfromtxt` because some paths through the code pass in other values to test how the converter behaves. For example, [this line](https://github.com/numpy/numpy/blob/1cf882248961a5d411c55d75bff190d7f7eb30e2/numpy/lib/_iotools.py#L804) passes in a `'1'` to each of the converters to test what `dtype` it outputs.

It is very confusing from the user side to see these extraneous `'1'`s passed in with no explanation when debugging, and then to see them not appear in the actual output array.
https://stackoverflow.com/questions/75783751/reading-csv-with-numpy-has-extra-row-of-1s

Furthermore, this behavior changes depending on what [exception the user raises](https://github.com/numpy/numpy/blob/1cf882248961a5d411c55d75bff190d7f7eb30e2/numpy/lib/_iotools.py#L805) and whether the user called `np.genfromtxt` with a [filling_value](https://github.com/numpy/numpy/blob/1cf882248961a5d411c55d75bff190d7f7eb30e2/numpy/lib/_iotools.py#L799).

### Idea or request for content:

The documentation around this behavior in `np.genfromtxt` should be more clear.

A few potential solutions:
- The user is not expected to throw exceptions in the converters
- The implementation could be changed to not pass in these testing values to the converters, or alternatively the user could be told under what circumstances they will be.

It would be problematic if a user has to write a converter with the logic "only accept `'A'` or `'B'`, except in the one situation where you might be passed an extraneous `'1'` by the framework on the first time".

### PR DESCRIPTION
**Pull request by Aryan-Mishra24**
Description
This PR updates the documentation for user-defined converters in np.genfromtxt, as requested in issue #23429. The changes include clarifications on the input and output types expected by converters, as well as examples of how to define and use converters in different scenarios.

Changes Made
Added a section on input and output types expected by converters.
Added examples of using converters with different input and output types.
Added a subsection on using converters with structured arrays.
Checklist
 Updated the documentation.
 Added new tests for the updated functionality.
 Passes all existing tests.
 Follows the NumPy code style.
 Updated the release notes, if necessary.
Related Issues
Closes #23429

### CODE REVIEW

**Comment by Aryan-Mishra24** [comment]
one was meant to be DOC: Update documentation for user-defined converters in np.genfromtxt and the other for copyright-policy-update

File: numpy/lib/_iotools.py [hunk 1]
```diff
@@ -790,8 +790,16 @@ def update(self, func, default=None, testing_value=None,
         `update` takes the same parameters as the constructor of
         `StringConverter`, except that `func` does not accept a `dtype`
         whereas `dtype_or_func` in the constructor does.
-
+
```
Comment (MatteoRaso) [inline numpy/lib/_iotools.py:793]:
```suggestion
```
No need for this blank line here.

File: numpy/lib/_iotools.py [hunk 2]
```diff
@@ -790,8 +790,16 @@ def update(self, func, default=None, testing_value=None,
...
         whereas `dtype_or_func` in the constructor does.
-
+        
+
+    Be aware that user-defined converters may receive extraneous inputs
+    during the type determination process. In particular, a '1' value will
+    be passed to the converter when determining the output dtype. This
+    behavior should be considered when writing custom converter functions,
+    and users should ensure that their converters can handle such input
+    without raising unexpected exceptions.
```
Comment (MatteoRaso) [inline numpy/lib/_iotools.py:800]:
```suggestion
        Be aware that user-defined converters may receive extraneous inputs
        during the type determination process. In particular, a '1' value will
        be passed to the converter when determining the output dtype. This
        behavior should be considered when writing custom converter functions,
        and users should ensure that their converters can handle such input
        without raising unexpected exceptions.
```
The indentation needs to match the indentation for the rest of the `notes` section.

**Review by MatteoRaso — CHANGES_REQUESTED** [review]
Thanks for the PR. A few suggestions to start.

**Comment by Aryan-Mishra24** [comment]
Thanks rookie mistake I know
Noted and won't be repeated

**Review by rossbar — COMMENTED** [review]
There is still an indentation problem here. Moreover I don't think this is in the right place - `StringConverter` is not intended for direct use in user code.

Since this has been stale for a while, I opened #23787 as a simple alternative.

**Comment by seberg** [comment]
Superseded by gh-23787 so closing.

**Pull request closed without merging**